File: baml_language/interface_probes/typescript/generated_concrete_methods.ts
/** Add concrete_facades.baml alongside the shared fixture; no network calls. */
import assert from 'node:assert/strict';
import * as b from './generated/baml_sdk/index.js';
import { BamlType, BamlTypeMap, setTypeMap } from '@boundaryml/baml-bridge';

const client = await b.vendor.openai.ResponsesClient.new_async({ model: 'facade-probe', api_key: 'unused' });
assert.ok(client instanceof b.vendor.openai.ResponsesClient);
assert.equal(await client.id(), 'openai/facade-probe');
const agent = await b.ai.Agent.new_async({ client });
assert.ok(agent instanceof b.ai.Agent);
agent.close();
client.close();

const box: b.FacadeBox<string> = await b.facade_box_async();
assert.ok(box instanceof b.FacadeBox);
assert.equal('value' in box, false);
assert.equal(await box.read(), 'Ada');
assert.equal(await box.replace('Grace'), 'Grace');
const echoed: number = await box.echo(9, { $types: { U: BamlType.from('int') } });
assert.equal(echoed, 9);
const same: b.FacadeBox<string> = await box.same();
assert.ok(same instanceof b.FacadeBox);
assert.equal(await same.replace('Lin', { marker: 1 }), 'Lin');
assert.equal(await box.read(), 'Lin');
const record = await b.facade_record_async(box);
assert.ok(record.value instanceof b.FacadeBox);

setTypeMap(new BamlTypeMap());
const copied = await box.echo(record, { $types: {
  U: b.FacadeRecordType(b.FacadeBoxType(BamlType.from('string'))),
} });
assert.ok(copied instanceof b.FacadeRecord);
assert.ok(copied.value instanceof b.FacadeBox);
assert.equal(await copied.value.read(), 'Lin');

const flavor: b.FacadeFlavor = await box.echo(b.FacadeFlavor.Vanilla, { $types: { U: b.FacadeFlavorType() } });
assert.equal(flavor, b.FacadeFlavor.Vanilla);
const collision: BamlType<b.EvidenceItem> = b.EvidenceItemType_();
void collision;
const clone = box.clone();
box.close();
const pending = clone.replace('Ada');
clone.close();
assert.equal(await pending, 'Ada');
assert.equal(await same.read(), 'Ada');
same.close();
record.value.close();
copied.value.close();
console.log('Generated generic concrete methods, checked Self results, clients and Agent factories passed');

File: baml_language/interface_probes/python/generated_concrete_methods.py
"""Execute generated concrete APIs, including class/default-method frames.

Generate the shared interfaces fixture with ../baml/concrete_facades.baml
added to baml_src, then run with that generated SDK on PYTHONPATH.
"""

from __future__ import annotations

import asyncio
import copy
import gc

import baml_sdk as sdk
from baml_sdk.ai import Agent
from baml_sdk.vendor.openai import ResponsesClient
from baml_bridge import proto
from baml_bridge.baml_py import _pending_transfer_count
from baml_bridge.typemap import BamlTypeMap, get_type_map, set_type_map


async def main() -> None:
    # Real stdlib factories and a concrete client input; no network request.
    client = await ResponsesClient.new_async(model="facade-probe", api_key="unused")
    assert type(client) is ResponsesClient
    assert await client.id() == "openai/facade-probe"
    agent = await Agent.new_async(client=client)
    assert type(agent) is Agent
    agent.close()
    client.close()

    greeter = await sdk.FriendlyGreeter.new_async(prefix="Hello")
    assert type(greeter) is sdk.FriendlyGreeter
    assert await greeter.greet("Ada") == "Hello, Ada!"
    assert await greeter.label() == "greeter"
    assert await sdk.welcome_async(greeter, "Ada") == "Hello, Ada!"
    assert not hasattr(greeter, "prefix")
    pending = greeter.greet("Grace")
    greeter.close()
    assert await pending == "Hello, Grace!"

    box = await sdk.facade_box_async()
    assert type(box) is sdk.FacadeBox
    assert not hasattr(box, "value")
    assert await box.read() == "Ada"
    assert await box.replace("Grace") == "Grace"
    assert await box.read() == "Grace"
    assert await box.echo(9, _types={"U": int}) == 9
    same = await box.same()
    assert type(same) is sdk.FacadeBox
    assert await same.replace("Lin", marker=1) == "Lin"
    assert await box.read() == "Lin"
    record = await sdk.facade_record_async(box)
    assert isinstance(record.value, sdk.FacadeBox)
    assert await record.value.read() == "Lin"

    saved = get_type_map()
    set_type_map(BamlTypeMap.from_lazy_entries({}, {}, {}))
    try:
        # A method's type tokens and record codecs come from its captured SDK.
        echoed = await box.echo(
            record, _types={"U": sdk.FacadeRecord[sdk.FacadeBox[str]]}
        )
        assert isinstance(echoed.value, sdk.FacadeBox)
        assert await echoed.value.read() == "Lin"
    finally:
        set_type_map(saved)

    try:
        await box.replace(9)
    except TypeError:
        pass
    else:
        raise AssertionError("wrong concrete type reached the body")
    assert await same.read() == "Lin"
    sibling = copy.copy(box)
    box.close()
    assert await sibling.read() == "Lin"
    assert await same.to_data_async() == {"value": "Lin"}
    assert (
        proto._ty_to_python_type(
            proto.python_type_to_wire_ty(sdk.FacadeBox[str]), saved
        )
        == sdk.FacadeBox[str]
    )
    sibling.close()
    same.close()
    record.value.close()
    echoed.value.close()
    gc.collect()
    assert _pending_transfer_count() == 0
    print("generated concrete facade methods, generic records and ownership passed")


if __name__ == "__main__":
    asyncio.run(main())
